Add fixes to products for when REPLAY IC's are used

EricSinsky-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
    Please use a short (<60 char), descriptive title above. It should complete the sentence "If merged, this PR will _____". Capitalize the first word and do not end with a period.

    PRs should meet these guidelines:
    - Each PR should address ONE topic and have an associated issue.
    - No hard-coded paths or personal directories.
    - No temporary or backup files should be committed (including logs).
    - Any code that you disabled by being commented out should be removed or reenabled.

    Please delete all these comments before submitting the PR.
-->
# Description
<!-- This description will become the commit message for the PR-->
<!-- Please use this format for your description:

  Describe your changes. Focus on the *what* and *why*. The *how* will be evident from the changes. In particular, be sure to note any interface changes, such as command line syntax, that will need to be communicated to users.

  At the end of your description, please be sure to add the issue this PR solves using the word "Resolves". If there are any issues that are related but not yet resolved (including in other repos), you may use "Refs".

  Resolves #1234
  Refs #4321
  Refs NOAA-EMC/repo#5678
-->

This PR fixes a couple issues that arise when replay initial conditions are used. These issues only occur when `REPLAY_ICS` is set to `YES` and `OFFSET_START_HOUR` is greater than `0`. The following items are addressed in this PR. 

1. Fix issue that causes ocean_prod tasks not to be triggered (issue [#2725](https://github.com/NOAA-EMC/global-workflow/issues/2725)). A new diag_table was added (called `diag_table_replay`) that is used only when `REPLAY_ICS` is set to `YES`. This diag_table accounts for the offset that occurs when using replay IC's. 
2. Fix issue that causes atmos_prod tasks not to be triggered for the first lead time (e.g. f003) (issue [#2754](https://github.com/NOAA-EMC/global-workflow/issues/2754)). When `OFFSET_START_HOUR` is greater than `0`, the first `fhr` is `${OFFSET_START_HOUR}+(${DELTIM}/3600)`, which is defined in `forecast_predet.sh` and will allow data for the first lead time to be generated. The filename with this lead time will still be labelled with `OFFSET_START_HOUR`. 
3. Minor modifications were made to the extractvars task so that atmos data from replay cases can be processed. 

This PR was split from PR #2680. 

Refs #2725, #2754 

# Type of change
<!-- Delete all except one -->
- Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on WCOSS
- Cycled test on Orion
- Forecast-only on Hera
-->

These changes were cloned, built and ran on WCOSS2. These changes were tested by running GEFS. 

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

EricSinsky-NOAA committed:
Update linters.yaml

The linters.yaml has been updated so that shellnorms can run whenever a push occurs to EricSinsky-NOAA/develop.

EricSinsky-NOAA committed:
Revert changes in linters.yaml

The changes to linters.yaml have been reverted.

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

EricSinsky-NOAA committed:
Add changes related to offset

Changes have been added to account for the fixes needed to be made when there is an offset. These changes involve adjusting the atmosphere and ocean fhrs.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

aerorahul commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -42,6 +46,18 @@ local SDAY=${current_cycle:6:2}
 
+if (( OFFSET_START_HOUR > 0 )); then
+  local current_cycle_p1 
+  current_cycle_p1=$(${NDATE} "${FHOUT_OCN}" "${current_cycle}")
+  local current_cycle_offset
+  current_cycle_offset=$(${NDATE} "${OFFSET_START_HOUR}" "${current_cycle}")
```
please use `date` instead of `NDATE`

EricSinsky-NOAA replied:
Updated.

aerorahul commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -17,6 +17,10 @@ if (( gwd_opt == 2 )) && [[ ${do_gsl_drag_ls_bl} == ".true." ]]; then
   cdmbgwd=${cdmbgwd_gsl}
 fi
 
+if (( OFFSET_START_HOUR != 0 )); then
+  DIAG_TABLE="${PARMgfs}/ufs/fv3/diag_table_replay"
+fi
```
Please set this in `config.fcst`

EricSinsky-NOAA replied:
This has been moved to `parm/config/gefs/config.fcst`.

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -204,8 +204,10 @@ EOF
     for fhr in ${FV3_OUTPUT_FH}; do
-      FH3=$(printf %03i "${fhr}")
-      FH2=$(printf %02i "${fhr}")
+      fhrf=$(echo "${fhr}" | bc -l)
+      fhri=$(printf %.0f "${fhrf}")
```
Can you explain what is the issue this seems to be solving?

EricSinsky-NOAA replied:
If `fhr` is a decimal number, it needs to be converted to a whole number here because we want to use a whole number in the filename of the output data. The first `fhr` will be a decimal number in cases where `OFFSET_START_HOUR > 0`. For example, if `OFFSET_START_HOUR` is `3`, the first `fhr` will need to be `${OFFSET_START_HOUR}+(${DELTIM}/3600)`, or 3.0833 if `DELTIM` is 300. If we keep `fhr` as 3, then output data for `f003` will not be generated.  I can add a brief comment to the script here.

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -451,11 +453,21 @@ MOM6_postdet() {
       # Native model output uses window midpoint in the filename, but we are mapping that to the end of the period for COM
-      source_file="ocn_${vdate_mid:0:4}_${vdate_mid:4:2}_${vdate_mid:6:2}_${vdate_mid:8:2}.nc"
-      dest_file="${RUN}.ocean.t${cyc}z.${interval}hr_avg.f${fhr3}.nc"
+      if (( OFFSET_START_HOUR > 0 )) &&  (( fhr == FHOUT_OCN ));then
+        source_file="ocn_lead1_${vdate_mid:0:4}_${vdate_mid:4:2}_${vdate_mid:6:2}_${vdate_mid:8:2}.nc"
+        dest_file="${RUN}.ocean.t${cyc}z.${interval}hr_avg.f${fhr3}.nc"
+      else
+        source_file="ocn_${vdate_mid:0:4}_${vdate_mid:4:2}_${vdate_mid:6:2}_${vdate_mid:8:2}.nc"
+        dest_file="${RUN}.ocean.t${cyc}z.${interval}hr_avg.f${fhr3}.nc"
+      fi
```
Since `dest_file` is the same, suggest removing it from the if-block.
```suggestion
      if (( OFFSET_START_HOUR > 0 )) &&  (( fhr == FHOUT_OCN ));then
        source_file="ocn_lead1_${vdate_mid:0:4}_${vdate_mid:4:2}_${vdate_mid:6:2}_${vdate_mid:8:2}.nc"
      else
        source_file="ocn_${vdate_mid:0:4}_${vdate_mid:4:2}_${vdate_mid:6:2}_${vdate_mid:8:2}.nc"
      fi
      dest_file="${RUN}.ocean.t${cyc}z.${interval}hr_avg.f${fhr3}.nc"
```

EricSinsky-NOAA replied:
Fixed.

aerorahul reviewed (commented):
See some suggestions.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Replace NDATE with date

NDATE has been removed and has been replaced with date.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Move where replay diag_table is defined

The determination on whether or not the replay_diag_table should be used has been moved to config.fcst.

EricSinsky-NOAA committed:
Move dest_file outside if-statement

The dest_file variable has been moved outside of if-statement in forecast_postdet.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Add bugfix in exglobal_stage_ic

A bugfix has been added in exglobal_stage_ic to ensure ${MEMDIR:3} is not intrepreted in base 8.

EricSinsky-NOAA committed:
Revert "Add bugfix in exglobal_stage_ic"

This reverts commit dcb8066ea5ed547c3a42bd04ae53a454da789e98.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Add fix for fv3 filename when fhr is decimal

A fix has been added for cases where there is an fhr that is a decimal number.

EricSinsky-NOAA committed:
Address shell check error

EricSinsky-NOAA committed:
Add more fixes for filename when fhr is decimal

EricSinsky-NOAA committed:
Set default REPLAY_ICS to NO in predet/postdet

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Address shell check errors

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Give extractvars ability to process replay cases

Modifications have been made to the extractvars task so that data from replay cases can be extracted.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Fix bug in config.resources for extractvars

A bug has been fixed in config.resources that originated from changes from PR#2727.

EricSinsky-NOAA commented on parm/config/gefs/config.resources:
```diff
@@ -272,7 +272,7 @@ case ${step} in
     export ntasks_gefs=1
     export threads_per_task_gefs=1
-    export tasks_per_node_gefs="${ntasks}"
+    export tasks_per_node_gefs="${ntasks_gefs}"
```
@DavidHuber-NOAA Adjustments needed to be made to the extractvars resources due to changes made from PR #2727. Could you let me know that this is correct?  I changed `ntasks` to `ntasks_gefs` because when `ntasks` was used here, I got the following error when setting up my experiment: 
```
  File "/lfs/h2/emc/ens/noscrub/eric.sinsky/GEFS/offsetfixes/workflow/rocoto/tasks.py", line 186, in get_resource
    nodes = int(np.ceil(float(ntasks) / float(ppn)))
ValueError: could not convert string to float: 
```

DavidHuber-NOAA replied:
@EricSinsky-NOAA Did you update your submodules locally (i.e. `git submodule update --init --recursive`)?  If this does not resolve the issue, can you point me to your experiment directory?

EricSinsky-NOAA replied:
I did update my submodules before setting up my experiment. My experiment can be set up successfully when I use `ntasks_gefs` instead of `ntasks`. Here is my experiment directory: `/lfs/h2/emc/ens/noscrub/eric.sinsky/GEFS/expdir/customexp/offsetfixes`.

DavidHuber-NOAA replied:
@EricSinsky-NOAA I apologize, I misread your initial post.  Yes, this looks correct.  Thanks for catching my bug.

EricSinsky-NOAA committed:
Modify config.extractvars based on requirements

Some parameters in config.extractvars have been modified based on reforecast requirements.

EricSinsky-NOAA commented:
A C48_S2SWA_gefs CI test case has successfully completed. Also tested a replay case and the ocean_prod jobs were triggered and the f003 atmos products were created. Marking as ready for review.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

aerorahul commented on parm/config/gefs/config.fcst:
```diff
@@ -242,7 +242,11 @@ export FSICS="0"
 #---------------------------------------------------------------------
 # Write more variables to output
-export DIAG_TABLE="${PARMgfs}/ufs/fv3/diag_table"
+if (( OFFSET_START_HOUR != 0 )); then
+  export DIAG_TABLE="${PARMgfs}/ufs/fv3/diag_table_replay"
+else
+  export DIAG_TABLE="${PARMgfs}/ufs/fv3/diag_table"
+fi
```
can this be conditioned on `REPLAY_ICS`?

EricSinsky-NOAA replied:
Sure. Done.

aerorahul commented on ush/parsing_namelists_FV3.sh:
```diff
@@ -42,6 +42,18 @@ local SDAY=${current_cycle:6:2}
 local CHOUR=${current_cycle:8:2}
 local MOM6_OUTPUT_DIR="./MOM6_OUTPUT"
 
+if (( OFFSET_START_HOUR > 0 )); then
```
Since this is only used for replay, should we change the conditional to `REPLAY_ICS`?

EricSinsky-NOAA replied:
Conditional has been changed to `REPLAY_ICS`.

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -145,6 +145,30 @@ FV3_predet(){
+  # Create an FV3 fhr list to be used in the namelist for when REPLAY_ICS is set to YES
+  # The FV3 fhr list for the namelist and the FV3 fhr list for the filenames
+  # are only different when REPLAY_ICS is set to YES
+  if [[ "${REPLAY_ICS:-NO}" == "YES"  ]]; then
+    FV3_OUTPUT_FH_NML="$(echo "scale=5; ${OFFSET_START_HOUR}+(${DELTIM}/3600)" | bc -l)"
+    FV3_OUTPUT_FH_s=$(( OFFSET_START_HOUR * 3600 + DELTIM ))
+    local FHMIN_REPLAY=$(( FHMIN + FHOUT ))
+    local fhr=${FHMIN_REPLAY}
+    if (( FHOUT_HF > 0 && FHMAX_HF > 0 )); then
+      FV3_OUTPUT_FH_NML="${FV3_OUTPUT_FH_NML} $(seq -s ' ' "${FHMIN_REPLAY}" "${FHOUT_HF}" "${FHMAX_HF}")"
+      FV3_OUTPUT_FH_s="${FV3_OUTPUT_FH_s} $(seq -s ' ' "$(( FHMIN_REPLAY * 3600 ))" "$(( FHOUT_HF * 3600 ))" "$(( FHMAX_HF * 3600 ))")"
+      fhr=${FHMAX_HF}
+    fi
+    FV3_OUTPUT_FH_NML="${FV3_OUTPUT_FH_NML} $(seq -s ' ' "${fhr}" "${FHOUT}" "${FHMAX}")"
+    FV3_OUTPUT_FH_s="${FV3_OUTPUT_FH_s} $(seq -s ' ' "$(( fhr * 3600 ))" "$(( FHOUT * 3600 ))" "$(( FHMAX * 3600 ))")"
+    local hh mm ss s_total
+    for s_total in ${FV3_OUTPUT_FH_s}; do
+      # Convert seconds to HHH:MM:SS
+      (( ss = s_total, mm = ss / 60, ss %= 60, hh = mm / 60, mm %= 60 )) || true
+      FV3_OUTPUT_FH_hhmmss+=("$(printf "%03d-%02d-%02d" "${hh}" "${mm}" "${ss}")")
+    done
+  fi
```
Why isn't there an `else` part for non-replay condition?

EricSinsky-NOAA replied:
I made some modifications here so that there is an `else` part.

aerorahul reviewed (commented):
A couple of minor comments.  The PR is ok otherwise.
I am just seeking feedback on the conditionals to be on `REPLAY_ICS` rather than `OFFSET_HOUR`.

EricSinsky-NOAA committed:
Condition based on REPLAY_ICS

A few if-blocks are conditioned based on REPLAY_ICS instead of OFFSET_START_HOUR.

EricSinsky-NOAA committed:
Rearange if-block to address reviewer comment

The if-block in forecast_predet has been rearranged so that it is more readable without compromising the general logic. Also FHMIN_REPLAY has been replaced by the already defined variable OFFSET_START_HOUR.

EricSinsky-NOAA committed:
Address shellcheck error

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -153,6 +154,32 @@ FV3_predet(){
+  # are only different when REPLAY_ICS is set to YES
+  if [[ "${REPLAY_ICS:-NO}" == "YES"  ]]; then
+    FV3_OUTPUT_FH_NML="$(echo "scale=5; ${OFFSET_START_HOUR}+(${DELTIM}/3600)" | bc -l)"
+    FV3_OUTPUT_FH_s=$(( OFFSET_START_HOUR * 3600 + DELTIM ))
```
This should be local.

EricSinsky-NOAA replied:
Changed this to local.

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -153,6 +154,32 @@ FV3_predet(){
+  # The FV3 fhr list for the namelist and the FV3 fhr list for the filenames
+  # are only different when REPLAY_ICS is set to YES
+  if [[ "${REPLAY_ICS:-NO}" == "YES"  ]]; then
+    FV3_OUTPUT_FH_NML="$(echo "scale=5; ${OFFSET_START_HOUR}+(${DELTIM}/3600)" | bc -l)"
```
Is there a reason to call this `FV3_OUTPUT_FH_NML`?  Why can't this just be `FV3_OUTPUT_FH`?

EricSinsky-NOAA replied:
When `REPLAY_ICS` is set to `YES`, the fhr used in the namelist (`FV3_OUTPUT_FH_NML`) will be different from the fhr list used in `FV3_postdet` (`FV3_OUTPUT_FH`). Th fhr list used in the namelist will have some decimal numbers (e.g. 3.08333, 6, 9, 12) and the fhr list used in `FV3_postdet` will have all whole numbers (e.g. 3, 6, 9, 12), which are converted to `FH3` and `FH2` then used in the filenames.

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -202,13 +202,23 @@ EOF
+      # When replaying, the time format outputted by model in filename is HH-MM-SS
+      # because first fhr is a decimal number
+      if [[ ${REPLAY_ICS:-NO} == "YES" ]] && (( fhr >= OFFSET_START_HOUR )); then
+        ${NLN} "${COMOUT_ATMOS_HISTORY}/${RUN}.t${cyc}z.atmf${FH3}.nc" "atmf${FV3_OUTPUT_FH_hhmmss[${iicnt}]}.nc"
```
How is this working? This seems to be an array defined in `forecast_predet.sh`, and arrays cannot be exported in bash!

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -153,6 +154,32 @@ FV3_predet(){
+    for s_total in ${FV3_OUTPUT_FH_s}; do
+      # Convert seconds to HHH:MM:SS
+      (( ss = s_total, mm = ss / 60, ss %= 60, hh = mm / 60, mm %= 60 )) || true
+      FV3_OUTPUT_FH_hhmmss+=("$(printf "%03d-%02d-%02d" "${hh}" "${mm}" "${ss}")")
```
Seems like this is being used in `forecast_postdet.sh`.  This is a bash array.  I have not been able to export bash arrays!  @WalterKolczynski-NOAA ??

WalterKolczynski-NOAA replied:
Explained this to Rahul before he went on leave. Both predet and postdet are called as functions in the main script, so the context never changes for these variables and they never need to be exported (which, indeed, would not work).

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Make variable FV3_OUTPUT_FH_s local

EricSinsky-NOAA committed:
Avoid exporting arrays in bash

EricSinsky-NOAA committed:
Fix potential bug

Fix a bug that would occur if REPLAY_ICS are used, an FHOUT of 6 is used and no FHOUT_HF_GFS is used.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Organize netcdf comout directories for ocniceprod

The Netcdf ocean and ice products have been separated into subdirectories based on product grid. This is consistent with how the Grib2 ocean and ice products are organized.

WalterKolczynski-NOAA reviewed (dismissed):
Conditionally approved

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

WalterKolczynski-NOAA commented:
Looks like PBS ate the job status of a bunch of jobs or something. I'll try again after the conflict is addressed:
```
202112210000                 gfsfcst                   147300374         UNAVAILABLE                   -         0           0.0
```

DavidHuber-NOAA commented:
I've seen this `UNAVAILABLE` status from `rocoto` after a `ruby` update (to v3+), though I don't think WCOSS2 is running a newer version.  `Rocoto` v1.3.7 resolved it for Orion and Hercules, but it does require updating `ruby`.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/offsetfixes

EricSinsky-NOAA committed:
Extract ice products with native grid instead

EricSinsky-NOAA commented:
@WalterKolczynski-NOAA The conflict has been resolved. This conflict was due to PR #2792, which was recently merged to develop. It appears PR #2792 adds the staging of the UFS cubed-sphere history files. I am having these files staged for both replay and non-replay cases. I have also added a minor change to the extractvars task, where I am extracting ice products with a native grid instead of a 0p25 grid.

EricSinsky-NOAA committed:
Remove unused local variable

The local variable iicnt has been removed from forecast_postdet.sh since it is no longer being used.

WalterKolczynski-NOAA commented:
> I've seen this `UNAVAILABLE` status from `rocoto` after a `ruby` update (to v3+), though I don't think WCOSS2 is running a newer version. `Rocoto` v1.3.7 resolved it for Orion and Hercules, but it does require updating `ruby`.

It also happens anytime the job info is no longer available from the scheduler. Usually because it was a long time between `rocotorun` calls or the scheduler/machine rebooted. Neither should've happened on WCOSS, which is why I went with the more nebulous 'scheduler ate it' ;)

EricSinsky-NOAA commented:
@WalterKolczynski-NOAA It looks like another stall happened on WCOSS2 due to an "UNAVAILABLE" state. The stall occurred for gdasprep. According to `rocotocheck`, all the dependencies were satisfied for gdasprep, so I am not sure what caused the task to be in an "UNAVAILABLE" state.

DavidHuber-NOAA commented:
@EricSinsky-NOAA @christopherwharrop-noaa @WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA
I have not seen this behavior on WCOSS2 before.  I wonder if `rocotorun` is periodically hanging.

The `gdasprep` job did submit successfully at 20:47:24, according to `/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/EXPDIR/C96_atmaerosnowDA_d443bf9c/logs/2021122018.log` and the job completed at 20:53:13 according to `/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/COMROOT/C96_atmaerosnowDA_d443bf9c/logs/2021122018/gdasprep.log`.  It is not clear to me why it did not pick up that the job completed successfully.  Unfortunately, it is too late to run `qstat -asx` on the JobID to find out what `PBS` thought the job did.

Checking the CI logs, I found this message in `/u/terry.mcguinness/ci_logs/run_ci_2024-08-08-20:55.log`, lines 463-466:
```
Running: /apps/ops/test/nco/core/rocoto.v1.3.5/bin/rocotorun -v 10 -w /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/EXPDIR/C96_atmaerosnowDA_d443bf9c/C96_atmaerosnowDA_d443bf9c.xml -d /lf
s/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/EXPDIR/C96_atmaerosnowDA_d443bf9c/C96_atmaerosnowDA_d443bf9c.db
+ run_ci.sh[87]/apps/ops/test/nco/core/rocoto.v1.3.5/bin/rocotorun -v 10 -w /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/EXPDIR/C96_atmaerosnowDA_d443bf9c/C96_atmaerosnowDA_d443bf9c.xml 
-d /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2755/RUNTESTS/EXPDIR/C96_atmaerosnowDA_d443bf9c/C96_atmaerosnowDA_d443bf9c.db
08/08/24 20:56:49 UTC :: C96_atmaerosnowDA_d443bf9c.xml :: Timeout::Error
```
I'm not sure what a timeout error means, but I suspect this is the culprit.  What do you think, @christopherwharrop-noaa?

DavidHuber-NOAA commented:
I also wonder if `rocotostat` and `rocotorun` are being called at the same time.  This is very hard to tell from the logs, but perhaps the `rocotostat` at line 121 of  `/u/terry.mcguinness/ci_logs/check_ci_2024-08-08-20:54.log` is still running when `rocotorun` is called in `run_ci_2024-08-08-20:55.log`.  Could that cause issues?

TerrenceMcGuinness-NOAA commented:
@DavidHuber-NOAA Christopher H. has a locking mech. for such race conditions on the db file so I would say that wouldn't be an issue directly.

UNAVAILABLE is usually indicative of issues with the scheduler BTW.  This means **Rocoto** has lost track of the job usually because of anomalies with the reporting from the scheduler.

I have isolated the issue and have been able to repeate the errors outside of the CI framework.
The timeout response occurred using Rocto with a fresh db file.
Followed by the `UNKNOWN `state being assigned to jobs.


```
terry.mcguinness (clogin02) C96_atmaerosnowDA_d443bf9c $ rocotostat -d C96_atmaerosnowDA_d443bf9c.db -w C96_atmaerosnowDA_d443bf9c.xml -s
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200      Active    Aug 09 2024 13:44:41             -          
202112201800      Active    Aug 09 2024 13:44:41             -          
202112210000      Active    Aug 09 2024 13:44:41             - 

         
terry.mcguinness (clogin02) C96_atmaerosnowDA_d443bf9c $ rocotostat -d C96_atmaerosnowDA_d443bf9c.db -w C96_atmaerosnowDA_d443bf9c.xml -c 202112201200
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202112201200                gdasfcst                   147657487             RUNNING                   -         0           0.0
202112201200     gdasatmos_prod_f000                   147657485           SUCCEEDED                   0         1          54.0
202112201200     gdasatmos_prod_f003                   147657486           SUCCEEDED                   0         1          51.0
202112201200     gdasatmos_prod_f006                   147657489           SUCCEEDED                   0         1          52.0
202112201200     gdasatmos_prod_f009                   147657488           SUCCEEDED                   0         1          53.0
terry.mcguinness (clogin02) C96_atmaerosnowDA_d443bf9c $ rocotorun -v 10 -w C96_atmaerosnowDA_d443bf9c.xml -d C96_atmaerosnowDA_d443bf9c.db
08/09/24 13:49:18 UTC :: C96_atmaerosnowDA_d443bf9c.xml :: Timeout::Error


terry.mcguinness (clogin02) C96_atmaerosnowDA_d443bf9c $ rocotostat -d C96_atmaerosnowDA_d443bf9c.db -w C96_atmaerosnowDA_d443bf9c.xml -c 202112201200                                                                                                             
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202112201200                gdasfcst                   147657487         UNAVAILABLE                   -         0           0.0
202112201200     gdasatmos_prod_f000                   147657485           SUCCEEDED                   0         1          54.0
202112201200     gdasatmos_prod_f003                   147657486           SUCCEEDED                   0         1          51.0
202112201200     gdasatmos_prod_f006                   147657489           SUCCEEDED                   0         1          52.0
202112201200     gdasatmos_prod_f009                   147657488           SUCCEEDED                   0         1          53.0
```

DavidHuber-NOAA commented:
Opened rocoto issue https://github.com/christopherwharrop/rocoto/issues/110.

It appears that the cause is a stalled call to `qstat`.  I'll report the stalled call to NCO as well.

christopherwharrop-noaa commented:
Rocoto has very little control over hanging batch system commands.  If a batch system is being hammered, and `qstat` starts taking a long time to run, all Rocoto can do is time out the commands and try again later.

I will say that PBSPro is highly prone to this type of behavior as it has had scaling/threading problems in the past.  Which is one reason why I was very surprised to see PBSPro selected as the batch system for WCOSS.  And it is very easy to overload PBSPro with naive user behaviors.  Rocoto has already been highly tuned to do everything possible to put as little load as possible on PBSPro as problems were first seen and addressed on Cheyenne.

christopherwharrop-noaa commented:
If you (or others) have other processes that run `qstat` commands on WCOSS, you may want to audit those to find out what they are doing.  PBSPro is very fragile, and someone running `qstat` or, even worse, `pbsnodes`, in the wrong way in automation could very easily cause problems.

TerrenceMcGuinness-NOAA commented:
The **enkfgdascleanup** job is  PENDING in the queue for the C96C48_hybatmDA case on Hera.
Not sure what **_launch failed requeued held_** means.
Not flagged as STALLED because the is still QUEUED when PENDING. 

```
Terry.McGuinness (hfe03) C96C48_hybatmDA_57107f48 $ squeue -u $USER
     JOBID PARTITION  NAME                     USER             STATE        TIME TIME_LIMIT NODES NODELIST(REASON)
  64649768 hera       C96C48_hybatmDA_57107f48 Terry.McGuinness PENDING      0:00      15:00     1 (launch failed requeued held)
  64649463 hera       C96_atmaerosnowDA_57107f Terry.McGuinness PENDING      0:00      30:00    10 (launch failed requeued held)
  
Terry.McGuinness (hfe03) C96C48_hybatmDA_57107f48 $ rocotostat -w C96C48_hybatmDA_57107f48.xml -d C96C48_hybatmDA_57107f48.db | grep QUE
202112210000         enkfgdascleanup                    64649768              QUEUED                   -         0           0.0

Terry.McGuinness (hfe03) C96C48_hybatmDA_57107f48 $ rocotocheck -w C96C48_hybatmDA_57107f48.xml -d C96C48_hybatmDA_57107f48.db -c 202112210000 -t enkfgdascleanup

Task: enkfgdascleanup
  account: nems
  command: /scratch1/NCEPDEV/global/CI/2755/gfs/jobs/rocoto/cleanup.sh
  cores: 1
  cycledefs: gdas
  final: false
  jobname: C96C48_hybatmDA_57107f48_enkfgdascleanup_00
  join: /scratch1/NCEPDEV/global/CI/2755/RUNTESTS/COMROOT/C96C48_hybatmDA_57107f48/logs/2021122100/enkfgdascleanup.log
  maxtries: 2
  memory: 4096M
  name: enkfgdascleanup
  nodes: 1:ppn=1:tpp=1
  partition: hera
  queue: batch
  throttle: 9999999
  walltime: 00:15:00
  environment
    CDATE ==> 2021122100
    COMROOT ==> /scratch1/NCEPDEV/global/CI/2755/RUNTESTS/COMROOT
    DATAROOT ==> /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C96C48_hybatmDA_57107f48/enkfgdas.2021122100
    EXPDIR ==> /scratch1/NCEPDEV/global/CI/2755/RUNTESTS/EXPDIR/C96C48_hybatmDA_57107f48
    HOMEgfs ==> /scratch1/NCEPDEV/global/CI/2755/gfs
    NET ==> gfs
    PDY ==> 20211221
    RUN ==> enkfgdas
    RUN_ENVIR ==> emc
    cyc ==> 00
  dependencies
    AND is satisfied
      SOME is satisfied
        enkfgdasearc00 of cycle 202112210000 is SUCCEEDED
        enkfgdasearc01 of cycle 202112210000 is SUCCEEDED

Cycle: 202112210000
  Valid for this task: YES
  State: active
  Activated: 2024-08-08 00:59:14 UTC
  Completed: -
  Expired: -

Job: 64649768
  State:  QUEUED (PENDING)
  Exit Status:  -
  Tries:  0
  Unknown count:  0
  Duration:  0.0
Terry.McGuinness (hfe03) C96C48_hybatmDA_57107f48 $ 
```

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA This looks like an issue with Slurm based on [this conversation](https://support.schedmd.com/show_bug.cgi?id=3535).  Here is an excerpt from Slurm support:
```
this usually means the launched failed on the compute nodes for one reason or another and instead of relaunching the job on the node and draining the queue it holds it to gain further review from the user or the admin.

What I would do is look at the slurmd log from one of the nodes where the job ran and see why the job failed.

A common mistake is the spool dir isn't owned by user root (or the slurmd isn't ran by user root).  But I am guessing this is a new thing and most other jobs have ran with no issue.

If you could look at the log and see if you can find something let me know.

If you can't easily see anything please attach the slurmd and slurmcltd logs and I can look and see if I can find something.

If you can also attach your slurm.conf file that would be helpful as well.
```

I'd suggest you report it to RDHPCS.

@DavidHuber-NOAA I will return to this shortly.  So far as I can tell (no log and no retires) that this job didn't run Still need to use Slurm queries on the slurm job number.

TerrenceMcGuinness-NOAA commented:
Made suggested updates to the Rocoto configurations for Timeouts and restated all the CI cases for this PR on WCOSS2:

```
terry.mcguinness (clogin03) 1.3.5 $ cat rocotorc 
---
:DatabaseType: SQLite3
:WorkflowDocType: XML
:DatabaseServer: true
:BatchQueueServer: true
:WorkflowIOServer: true
:MaxUnknowns: 3
:MaxLogDays: 7
:AutoVacuum: true
:VacuumPurgeDays: 30
:SubmitThreads: 8
:JobQueueTimeout: 120
:JobAcctTimeout: 120

terry.mcguinness (clogin03) 1.3.5 $ crontab -l
MAILTO="terry.mcguinness@noaa.gov>"
SHELL=/bin/bash -l
ci_dir=/lfs/h2/emc/global/noscrub/terry.mcguinness/GW/global-workflow_ci/ci/scripts
d_cmd=date +%Y-%m-%d-%H:%M
*/4 * * * * d=$($d_cmd); ${ci_dir}/driver.sh >& ~/ci_logs/bash_driver_${d}.log || echo "ERROR in driver"
*/7 * * * * d=$($d_cmd); ${ci_dir}/run_ci.sh >& ~/ci_logs/run_ci_${d}.log || echo "ERROR in run_ci"
*/6 * * * * d=$($d_cmd); ${ci_dir}/check_ci.sh >& ~/ci_logs/check_ci_${d}.log || echo "ERROR in check_ci"

terry.mcguinness (clogin03) 1.3.5 $ q

cbqs01: 
                                                                 Req'd  Req'd   Elap
Job ID               Username Queue    Jobname    SessID NDS TSK Memory Time  S Time
-------------------- -------- -------- ---------- ------ --- --- ------ ----- - -----
148101461.cbqs01     terry.m* dev      C48_S2SW_* 129237   1 128    --  03:00 R 00:04
148101541.cbqs01     terry.m* dev      C96_atmae* 172194   1 128    --  00:20 R 00:02
148101616.cbqs01     terry.m* dev      C96C48_hy* 117488   4 500    --  01:20 R 00:02
148101617.cbqs01     terry.m* dev      C96C48_hy* 187580   1  80    --  00:15 R 00:02
148101618.cbqs01     terry.m* dev      C96C48_hy*  70962   4 500    --  01:00 R 00:02
148101695.cbqs01     terry.m* dev      C96C48_uf*  67337   1 128    --  00:20 R 00:01
148101706.cbqs01     terry.m* dev      C96C48_uf*  73148   1 128    --  00:20 R 00:01
148101707.cbqs01     terry.m* dev      C96C48_uf* 212313   1 128    --  00:20 R 00:01

terry.mcguinness (clogin03) 1.3.5 $ displaydb
2755 Open Running 0 ci_repo
terry.mcguinness (clogin03) 1.3.5 $ 
```

TerrenceMcGuinness-NOAA commented:
**Confirmed:** The STALLED condition on WCOSS2 is a false negative. No UNAVAILBLE were observed, only UNKOWNS.   Added tasking to make the STALLED flag more robust.

TerrenceMcGuinness-NOAA commented:
Got passed PBS/Rocoto anomalies that were leading to false negatives for STALLING because of UNKNOWN/UNAVAILBLE states,  and arrived at `gsi.x` fails.  Starting CI up one more time to capture results.

KateFriedman-NOAA commented:
@EricSinsky-NOAA FYI, I have a PR in CI testing that refactors the staging job: https://github.com/NOAA-EMC/global-workflow/pull/2651

I ended up removing the `DTG_PREFIX` that is built using cycle and `OFFSET_START_HOUR` and replaced with `model_start_date_current_cycle` and associated logic to set that variable. I think my PR may add conflicts with yours if it goes in first. If you have a moment, please review the staging job PR and let me know if we'll need to add the usage of `OFFSET_START_HOUR` back in and update the `REPLAY_ICS` blocks in the new yaml (see links below). I was not able to test with `REPLAY_ICS=YES` for my branch testing.

https://github.com/KateFriedman-NOAA/global-workflow/blob/feature/issue_2475/parm/stage/stage.yaml.j2#L104
https://github.com/KateFriedman-NOAA/global-workflow/blob/feature/issue_2475/parm/stage/stage.yaml.j2#L166

If my PR goes in first, I can work with you to make any needed updates in your branch to accommodate the staging job refactor. Let me know!

WalterKolczynski-NOAA commented:
Since (a) the previous failures on WCOSS were unrelated to this PR and (b) everything changes is covered by tests run on other machines, I'm going to go ahead and merge this.

FYI: @KateFriedman-NOAA

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/offsetfixes

EricSinsky-NOAA commented:
> @EricSinsky-NOAA FYI, I have a PR in CI testing that refactors the staging job: #2651
> 
> I ended up removing the `DTG_PREFIX` that is built using cycle and `OFFSET_START_HOUR` and replaced with `model_start_date_current_cycle` and associated logic to set that variable. I think my PR may add conflicts with yours if it goes in first. If you have a moment, please review the staging job PR and let me know if we'll need to add the usage of `OFFSET_START_HOUR` back in and update the `REPLAY_ICS` blocks in the new yaml (see links below). I was not able to test with `REPLAY_ICS=YES` for my branch testing.
> 
> https://github.com/KateFriedman-NOAA/global-workflow/blob/feature/issue_2475/parm/stage/stage.yaml.j2#L104 https://github.com/KateFriedman-NOAA/global-workflow/blob/feature/issue_2475/parm/stage/stage.yaml.j2#L166
> 
> If my PR goes in first, I can work with you to make any needed updates in your branch to accommodate the staging job refactor. Let me know!

@KateFriedman-NOAA Thank you for letting us know about these conflicts with the replay-related variables. We can continue working to resolve these conflicts in @NeilBarton-NOAA's PR #2788.
